pipelining_example/pipelining_example.c

```diff
@@ -0,0 +1,2793 @@
+	struct timeval cur_time;
+	struct ibv_wc wc;
+	int poll_result;
+	int rc = 0;
+
+	/* poll the completion for a while before giving up of doing it .. */
+	gettimeofday(&cur_time, NULL);
+	start_time_msec = (cur_time.tv_sec * 1000) + (cur_time.tv_usec / 1000);
+	do {
+		poll_result = ibv_poll_cq(res->cq, 1, &wc);
+		gettimeofday(&cur_time, NULL);
+		cur_time_msec =
+		    (cur_time.tv_sec * 1000) + (cur_time.tv_usec / 1000);
+
+		server_check_async_event(res);
+
+	} while ((poll_result == 0) &&
+		 ((cur_time_msec - start_time_msec) < MAX_POLL_CQ_TIMEOUT));
+	if (poll_result < 0) {
+		/* poll CQ failed */
+		fprintf(stderr, "poll CQ failed\n");
+		rc = 1;
+	} else if (poll_result == 0) {
+		/* the CQ is empty */
+		fprintf(stderr,
+			"completion wasn't found in the CQ after timeout\n");
+		rc = 1;
+	} else {
+		/* CQE found */
+		/** fprintf( */
+		/**     stdout, */
+		/**     "completion was found in CQ with status 0x%x, opcode %u\n", */
+		/**     wc.status, wc.opcode); */
+		/* check the completion status (here we don't care about the
+		 * completion
+		 * opcode */
+		if (wc.status != IBV_WC_SUCCESS) {
+			fprintf(stderr, "got bad completion with status: 0x%x, "
+					"vendor syndrome: 0x%x\n",
+				wc.status, wc.vendor_err);
+			rc = 1;
+		}
+	}
+	return rc;
+}
+
+static inline void init_send_wr(struct resources *res, uint64_t req_id,
+				uint64_t send_flags)
+{
+	struct tx_desc *desc = get_tx_desc(res, req_id);
+	struct ibv_send_wr *wr = next_send_wr(desc);
+
+	memset(wr, 0, sizeof(*wr));
+	wr->opcode = IBV_WR_SEND;
+	wr->send_flags = send_flags;
+	wr->sg_list = &desc->msg_sge;
+	wr->num_sge = 1;
+	wr->wr_id = req_id;
+}
+
+static inline int post_send(struct ibv_qp *qp, struct tx_desc *desc)
+{
+	struct ibv_send_wr *bad_wr;
+	int rc;
+
+	rc = ibv_post_send(qp, first_send_wr(desc), &bad_wr);
+	desc->next_wr_idx = 0;
+
+	return rc;
+}
+
+int configure_sig_mkey(struct resources *res,
+		       struct mlx5dv_sig_block_attr *sig_attr,
+		       struct task *task) {
+	struct ibv_qp *qp;
+	struct ibv_qp_ex *qpx;
+	struct mlx5dv_qp_ex *dv_qp;
+	struct mlx5dv_mkey *mkey;
+	uint64_t data_addr;
+	struct repeated_block *rpb;
+	struct tx_desc *desc;
+	struct mlx5dv_mr_interleaved mr_interleaved[2];
+	uint32_t access_flags = IBV_ACCESS_LOCAL_WRITE |
+				IBV_ACCESS_REMOTE_READ |
+				IBV_ACCESS_REMOTE_WRITE;
+
+	desc = get_tx_desc(res, task->req_id);
+
+	mkey = desc->sig_mr;
+
+	qp = get_qp(res);
+	qpx = ibv_qp_to_qp_ex(qp);
+	dv_qp = mlx5dv_qp_ex_from_ibv_qp_ex(qpx);
+
+	ibv_wr_start(qpx);
+	qpx->wr_id = 0;
+	qpx->wr_flags = IBV_SEND_SIGNALED | IBV_SEND_INLINE;
+
+	mlx5dv_wr_mkey_configure(dv_qp, mkey, 0);
+	mlx5dv_wr_set_mkey_access_flags(dv_qp, access_flags);
+
+	rpb = &task->repeated_block;
+
+	data_addr = get_ind_buffer_addr(&task->data_buf);
+
+	/* data */
+	mr_interleaved[0].addr = data_addr;
+	mr_interleaved[0].bytes_count = rpb->data_size;
+	mr_interleaved[0].bytes_skip = 0;
+	mr_interleaved[0].lkey = get_ind_buffer_lkey(&task->data_buf);
```

Comment:
You have a continuous buffer. It does not make sense to use mlx5dv_wr_set_mkey_layout_interleaved() in this case. It is enough to use a one ibv_sge and mlx5dv_wr_set_mkey_layout_list(). I think the following sge does the same as the above interleaved:

sge.addr = data_addr;
sge.length = rpb->data_size * SERVER_MAX_BLOCKS_PER_IO;
sge.lkey = get_ind_buffer_lkey(&task->data_buf);

And the sge looks cleaner.

Reply:
Ok, I will change it, thanks!